Initial implementation of date picker feature

dmytro-palii opened the pull request.

paliid committed:
Initial implementation of date picker feature

alexromanov commented on .idea/dbnavigator.xml:
Do not include it in PR.

dmytro-palii replied:
Done

alexromanov commented on .idea/.gitignore:
Exclude .idea from using global gitignore

dmytro-palii replied:
Done

alexromanov commented on enums/__pycache__/global_enums.cpython-312.pyc:
Exclude it from PR

dmytro-palii replied:
Done

alexromanov commented on tests/__pycache__/conftest.cpython-312-pytest-8.3.4.pyc:
Do not include __pycache__ folder

dmytro-palii replied:
done

alexromanov commented on tests/conftest.py:
```diff
@@ -1,23 +1,57 @@
+
 @pytest.fixture(scope="session")
 def browser(playwright_instance):
+    # browser = playwright_instance.chromium.launch(headless=False)
```
Do not include commented code

dmytro-palii replied:
Done/Explanation

alexromanov replied:
Add custom parameter for pytest run to indicate browser mode. By default it should false. It true - show browser.

dmytro-palii replied:
Done

alexromanov commented on tests/conftest.py:
```diff
@@ -1,23 +1,57 @@
+
+@pytest.fixture(scope='module')
+def test_validation_month_day_year_data():
+    with open('validation_month_day_year_data.json') as file:
```
make it a single util method for reading json content

dmytro-palii replied:
Done

alexromanov commented on tests/test_date_picker.py:
```diff
@@ -0,0 +1,81 @@
+logging.basicConfig(level=logging.DEBUG)
+
+
+# Functional Test Cases
```
Split it into set of different test cases for datepicker

dmytro-palii replied:
Question

alexromanov commented on tests/test_date_picker_box.py:
Do not include empty py files into PR

dmytro-palii replied:
Done

alexromanov commented on tests/test_inputs.py:
```diff
@@ -0,0 +1,52 @@
+    that the input field's value matches the expected value.
+
+    Parameters:
+    page : Page
```
If you prefer to describe the test in the comments - do it, but do not include description of parameters (specifically pages, etc.

dmytro-palii replied:
Question

alexromanov commented on tests/test_inputs.py:
```diff
@@ -0,0 +1,52 @@
+
+
+@pytest.mark.smoke
+@pytest.mark.parametrize('text_input_value', ['Some text, with numbers 123!'])
```
It is better not to use parameterization with one value.

dmytro-palii replied:
Question

alexromanov commented on tests/test_scroll_list.py:
```diff
@@ -0,0 +1,47 @@
+
+
+@pytest.mark.smoke
+@pytest.mark.parametrize('list_option_value, list_text', [
```
Not in the current case, but use meaningful names for variables, even in parameterization.

dmytro-palii replied:
Question

alexromanov commented on tests/test_scroll_list.py:
```diff
@@ -0,0 +1,47 @@
+    assert scroll_list.input_value() == list_option_value, f'Selected value is not {list_option_value}'
+
+    selected_text = scroll_list.locator('option:checked').text_content()
+    assert ''.join(selected_text.split()) == list_text, F'Displayed text is not {list_text}'
```
Use lowercase f for f strings.

dmytro-palii replied:
Done

alexromanov commented on tests/test_scroll_list.py:
```diff
@@ -0,0 +1,47 @@
+    assert sorted_list.input_value() == list_option_value, f'Selected value is not {list_option_value}'
+
+    selected_text = sorted_list.locator('option:checked').text_content()
+    assert ''.join(selected_text.split()) == list_text, F'Displayed text is not {list_text}'
```
test_scroll_list_color() and test_scroll_list_sorted() can be extracted into one method with parameters

dmytro-palii replied:
Done

alexromanov commented on .gitignore:
Use common gitignore sample for python projects

dmytro-palii replied:
Question

alexromanov commented on tests/test_txt_date_picker.py:
```diff
@@ -0,0 +1,202 @@
+    ('10', 'Nov'),
+    ('11', 'Dec')
+])
+def test_months_select(page, list_option_value, list_text):
```
use more descriptive names for tests if you don't have additional explanation of test logic in the comments.

dmytro-palii replied:
Done

alexromanov commented on tests/test_txt_date_picker.py:
```diff
@@ -0,0 +1,202 @@
+    ('2034', '2034'),
+    ('2035', '2035')
+])
+def test_years_select(page, list_option_value, list_text):
```
Can we extract test_months_select() and test_years_select() into a common method with parameters?

dmytro-palii replied:
Done

alexromanov commented on utils/helpers.py:
```diff
@@ -0,0 +1,43 @@
+
+
+def month_name_to_index(month_name):
+    months = {
```
You can use single array and get month by index. 
Also, you can either have two array with full and small month names or make custom data structure to make it unified.

dmytro-palii replied:
Question

alexromanov replied:
You can create one list with week_day_full_names and week_day_short_names and get values by index.

alexromanov replied:
E.g., `week_day_short_names[0]` will be equal to "Jan"

alexromanov replied:
E.g, `week_day_short_names.index("Jan")` will return 0

alexromanov commented on utils/collect_validation_month_day_year_data.py:
```diff
@@ -0,0 +1,54 @@
+            page.locator('.ui-datepicker-month').select_option(str(month_index))
+            dates = page.locator('.ui-datepicker-calendar td[data-handler="selectDay"] a').all_inner_texts()
+
+            # actual_dates = [
```
Do not include commented code.

dmytro-palii replied:
Done

alexromanov commented on tests/test_txt_date_picker.py:
```diff
@@ -0,0 +1,202 @@
+    # Index for keeping track of the expected (year, month) during validation
+    index = expected_dates.index((start_year, start_moth))
+
+    for click_next in range(24):
```
Do not use "magic numbers" in the code. Make a constant.

dmytro-palii replied:
Question

alexromanov commented on tests/test_txt_date_picker.py:
```diff
@@ -0,0 +1,202 @@
+
+@pytest.mark.smoke
+def test_next_month_year(page, test_run_month_day_year_data):
+    txt_date = page.locator('#txtDate')
```
test_prev_month_year() and test_next_month_year() can be extracted into one common method

dmytro-palii replied:
Question

alexromanov replied:
Extract whatever common things you can into separate method and use it in both test cases.

alexromanov commented on tests/test_txt_date_picker.py:
```diff
@@ -0,0 +1,202 @@
+        current_month = month_locator.locator('option[selected]').text_content()
+
+        expected_year, expected_month = expected_dates[index]
+        print(f"After clicking next: Current year: {current_year}, Current month: {current_month}")
```
Use logger instead of print

dmytro-palii replied:
Done

alexromanov commented on tests/test_txt_date_picker.py:
```diff
@@ -0,0 +1,202 @@
+    min_year = current_year - 0
+    max_year = current_year + 0
+
+    for year, months in test_run_month_day_year_data.items():
```
Rather than having complex logic in one test - let's think of how we can split it into multiple tests.

dmytro-palii replied:
Question

paliid committed:
Stop tracking .idea/dbnavigator.xml

paliid committed:
Stop tracking enums/__pycache__/global_enums.cpython-312.pyc

paliid committed:
Removed .idea/ from tracking

paliid committed:
"Removed .idea/ from tracking"

paliid committed:
Update PR excluding ignored files

paliid committed:
Removed __pycache__ from tracking

alexromanov commented on tests/test_date_picker_box.py:
```diff
@@ -0,0 +1,117 @@
+def test_hover_date_picker_box(page):
+    """
+    @brief Verifies that hovering over the date picker box changes its size.
+    @param page The Playwright page instance.
```
We can skip describing page input.

alexromanov commented on tests/test_date_picker_box.py:
```diff
@@ -0,0 +1,117 @@
+    4. Capture the width and height after hovering.
+    5. Verify that the width and height change upon hovering.
+
+    @assert The width of the `.date-picker-box` should change after hovering.
```
We can skip describing asserts if we described all steps before.

alexromanov commented on tests/test_date_picker_box.py:
```diff
@@ -0,0 +1,117 @@
+
+    initial_color = submit_button.evaluate("element => window.getComputedStyle(element).getPropertyValue('background-color')")
+    submit_button.hover()
+    page.wait_for_timeout(200)
```
Move any magic numbers into constants. E.g., wait timeout.

alexromanov commented on tests/test_date_picker_box.py:
```diff
@@ -0,0 +1,117 @@
+    @assert The displayed date range must match the expected value.
+    """
+
+    # Use the fixture-generated dates
```
We can remove comments if you have clearly defined steps. Just keep the structure as you have now.

alexromanov commented on tests/test_inputs.py:
```diff
@@ -0,0 +1,38 @@
+def test_name_input(page, name_input_value):
+    name_input = page.locator("#name")
+    name_input.fill(name_input_value)
+    assert name_input.input_value() == "Test", f"Actual value is not equal to {name_input_value}"
```
Use faker library to generate name, email, phone number in tests.

alexromanov commented on tests/test_inputs.py:
```diff
@@ -0,0 +1,38 @@
+
+
+@pytest.mark.smoke
+@pytest.mark.parametrize("email_input_value", ["test@testson@mail.com"])
```
Use parameterize with list of correct / incorrect emails or phone numbers.

alexromanov commented on tests/test_static_web_table.py:
```diff
@@ -0,0 +1,50 @@
+        row_list = [cell for cell in row.locator('td').all_inner_texts()]
+        actual_rows_data_list.append(row_list)
+
+    missing_rows = [row for row in expected_rows_data_list if row not in actual_rows_data_list]
```
Can be done as operations with sets data structure

alexromanov commented on tests/test_txt_date_picker.py:
```diff
@@ -0,0 +1,244 @@
+    txt_date = page.locator('#txtDate')
+    txt_date.click()
+
+    date_picker_selector = page.locator('.ui-datepicker-month') if select_type == 'month' else page.locator('.ui-datepicker-year')
```
Probably you can use match expression here

alexromanov commented on tests/test_txt_date_picker.py:
```diff
@@ -0,0 +1,244 @@
+    @brief Verifies that selecting a date from the date picker inputs the correct value.
+
+    @param page The Playwright page instance.
+    @param read_file The fixture that reads JSON test data files.
```
json_test_data

alexromanov replied:
As an improvement we can implement general purpose fixture for reading json data, and add two fixtures `run_month_day_year_data` and `validation_month_day_year_data` which will use this general fixture inside. 

Then the test should have two particular fixtures as an inputs instead of general one.

alexromanov commented on tests/test_txt_date_picker.py:
```diff
@@ -0,0 +1,244 @@
+
+    txt_date = page.locator('#txtDate')
+
+    current_year = datetime.now().year
```
Extra algorithm into separate method with current_year, min_year and max_year as a parameters and use it in test

alexromanov commented on tests/test_upload_files.py:
```diff
@@ -0,0 +1,74 @@
+    """
+
+    file_input = page.locator('input#multipleFilesInput')
+    file_input.set_input_files(file_path)
```
Why do you have this line, if you are setting input files at the line 56?

alexromanov commented on utils/collect_pagination_web_table.py:
```diff
@@ -0,0 +1,56 @@
+    pagination = page.locator("#pagination a")
+    for page in range(pagination.count()):
+        pagination.nth(page).click()
+        time.sleep(2)
```
Extract number into constant. And why have you used time.sleep here?

alexromanov commented on utils/collect_test_run_month_day_year_data.py:
```diff
@@ -0,0 +1,51 @@
+    return all_dates_dict
+
+
+def save_data_to_file(data, file_name='test_run_month_day_year_data.json'):
```
Extract save_data_to_file from all utils into separate method or even class

alexromanov commented on utils/collect_validation_month_day_year_data.py:
```diff
@@ -0,0 +1,77 @@
+from datetime import datetime
+
+
+def collect_validation_month_day_year_data(page):
```
Check if both of methods should exist: the current one and `collect_test_run_month_day_year_data`

alexromanov replied:
If possible - make one method and use it in both for months and for years.

alexromanov commented on utils/helpers.py:
```diff
@@ -0,0 +1,110 @@
+def day_month_format(date_line):
+    date_list = date_line.split('/')
+    day = date_list[0]
+    month = date_list[1]
```
`day, month, yeat = date_list`

alexromanov commented on utils/helpers.py:
```diff
@@ -0,0 +1,110 @@
+    month = date_list[1]
+    year = date_list[2]
+    for date in date_list:
+        if len(date) == 1:
```
`if date is not empty`

alexromanov commented on utils/helpers.py:
```diff
@@ -0,0 +1,110 @@
+    return date_string.split()[-2]
+
+
+def generate_start_end_date():
```
@alexromanov Take a look at logic.

alexromanov commented on utils/helpers.py:
```diff
@@ -0,0 +1,110 @@
+    start = datetime.strptime(start_date, '%Y-%m-%d')
+    end = datetime.strptime(end_date, '%Y-%m-%d')
+
+    difference = abs((start - end).days)
```
`return  abs((start - end).days)`

paliid committed:
Updated tests and utils, added pagination data file

paliid committed:
feat: Enhance test suite with new tests, improved documentation, and refactoring
This commit introduces significant enhancements to the test suite, including new test cases, improved documentation, and general code cleanup.
- **New Test Cases:**
    - Added comprehensive tests for date picker functionality in 	est_date_picker_one.py and 	est_date_picker_two.py.
    - Added new tests to 	est_date_picker_three.py to verify hover effects and date range submission.
    - Implemented tests for single and multiple file uploads in 	est_upload_files.py.
- **Documentation Improvements:**
    - Corrected the inaccurate docstrings for 	est_name_input and 	est_email_input in 	est_inputs.py to accurately reflect the test logic.
    - Added docstrings to 	est_phone_input and 	est_textarea_input in 	est_inputs.py for improved clarity.
- **Code Refactoring and Cleanup:**
    - Removed the unused, commented-out 
ead_file fixture in conftest.py.
    - Fixed a bug in 	est_textarea_input where the wrong variable (	extarea_input) was used in the assertion; it now correctly uses 	ext_input_value.
- **General Improvements:**
    - Implemented new tests, and improve existing tests and documentation
    - Refactor tests code and improve code quality
    - General refactor and code cleanup.
These changes significantly improve the robustness, clarity, and maintainability of the test suite.

paliid committed:
-   Added `pagination_web_table_validation_data.json` to store expected data for each page of the table.
